File: backend/src/app/routers/quest.py
from __future__ import annotations

import uuid
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_session
from app.models.quest import DailyQuest, QuestStatus, QuizAttempt
from app.routers.deps import current_user_id
from app.schemas.quest import AnswerIn, AnswerOut, PracticeOut, QuestOut, QuizQuestionOut
from app.services import memory_service
from app.services.quest_generator import ensure_today_quest

router = APIRouter(prefix="/quest", tags=["quest"])


def _to_out(q: DailyQuest) -> QuestOut:
    return QuestOut(
        id=q.id,
        date=q.date,
        concept_id=q.concept_id,
        explanation=q.explanation,
        # Strip answer_index/explanation from the response — clients learn nothing.
        quiz=[QuizQuestionOut(question=item["question"], options=item["options"]) for item in q.quiz_payload],
        practice=PracticeOut(**q.practice_payload) if q.practice_payload else PracticeOut(type="thought", prompt=""),
        status=q.status,
        completed_at=q.completed_at,
    )


@router.get("/today", response_model=QuestOut)
async def get_today(
    user_id: uuid.UUID = Depends(current_user_id),
    session: AsyncSession = Depends(get_session),
) -> QuestOut:
    quest = await ensure_today_quest(session, user_id)
    if quest is None:
        raise HTTPException(503, "knowledge graph empty or no candidate concept available")
    await session.commit()
    return _to_out(quest)


@router.post("/{quest_id}/answer", response_model=AnswerOut)
async def submit_answer(
    quest_id: uuid.UUID,
    body: AnswerIn,
    user_id: uuid.UUID = Depends(current_user_id),
    session: AsyncSession = Depends(get_session),
) -> AnswerOut:
    quest = await session.get(DailyQuest, quest_id)
    if quest is None or quest.user_id != user_id:
        raise HTTPException(404, "quest not found")
    if body.question_index < 0 or body.question_index >= len(quest.quiz_payload):
        raise HTTPException(400, "question_index out of range")

    question = quest.quiz_payload[body.question_index]
    is_correct = body.choice_index == question["answer_index"]

    session.add(
        QuizAttempt(
            quest_id=quest.id,
            question_index=body.question_index,
            user_answer=str(body.choice_index),
            is_correct=is_correct,
        )
    )
    await memory_service.record_answer(session, user_id, quest.concept_id, is_correct)

    # When all questions are answered, mark Quest complete.
    answered_count = body.question_index + 1  # rough proxy; refine with a real count query later
    if answered_count >= len(quest.quiz_payload):
        quest.status = QuestStatus.completed.value
        quest.completed_at = datetime.now(UTC)

    await session.commit()
    return AnswerOut(
        is_correct=is_correct,
        correct_index=question["answer_index"],
        explanation=question.get("explanation", ""),
    )

File: mobile/lib/models/quest.dart
class QuizQuestion {
  QuizQuestion({required this.question, required this.options});

  final String question;
  final List<String> options;

  factory QuizQuestion.fromJson(Map<String, dynamic> j) => QuizQuestion(
        question: j['question'] as String,
        options: (j['options'] as List).cast<String>(),
      );
}

class Practice {
  Practice({required this.type, required this.prompt});

  final String type;
  final String prompt;

  factory Practice.fromJson(Map<String, dynamic> j) => Practice(
        type: j['type'] as String? ?? 'thought',
        prompt: j['prompt'] as String? ?? '',
      );
}

class Quest {
  Quest({
    required this.id,
    required this.date,
    required this.conceptId,
    required this.explanation,
    required this.quiz,
    required this.practice,
    required this.status,
  });

  final String id;
  final DateTime date;
  final String conceptId;
  final String explanation;
  final List<QuizQuestion> quiz;
  final Practice practice;
  final String status;

  factory Quest.fromJson(Map<String, dynamic> j) => Quest(
        id: j['id'] as String,
        date: DateTime.parse(j['date'] as String),
        conceptId: j['concept_id'] as String,
        explanation: j['explanation'] as String? ?? '',
        quiz: (j['quiz'] as List)
            .map((e) => QuizQuestion.fromJson(e as Map<String, dynamic>))
            .toList(),
        practice: Practice.fromJson(j['practice'] as Map<String, dynamic>),
        status: j['status'] as String,
      );
}

class AnswerResult {
  AnswerResult({required this.isCorrect, required this.correctIndex, required this.explanation});

  final bool isCorrect;
  final int correctIndex;
  final String explanation;

  factory AnswerResult.fromJson(Map<String, dynamic> j) => AnswerResult(
        isCorrect: j['is_correct'] as bool,
        correctIndex: j['correct_index'] as int,
        explanation: j['explanation'] as String? ?? '',
      );
}
